### PR TITLE
feat(codex-pair): Stop-gate — opt-in `blockOn: HIGH` turn-end gate (#142, ADR-118)

### LINKED ISSUE #142: enhancement: codex-pair is advisory-only — add a Stop-hook gate (+ ack/ledger) so HIGH findings cannot be silently ignored
## Summary

codex-pair's per-edit findings are **advisory-only** (stderr → "read next turn") with **no enforcement**. In a long session this means HIGH findings can be silently ignored, and per-edit review floods `log.jsonl` with transient mid-edit findings that drown the real persistent ones. Proposal: add a **`Stop`-hook gate** that blocks the agent from ending a turn while unaddressed HIGH findings remain on files changed *this turn* — reusing the reviews already in `log.jsonl` (no extra Codex calls).

## Lived experience (field repro, ADR-095-style)

During a dense multi-hour refactor session (React Native, ~40 edits), codex-pair fired on ~30 edits and produced **~250 HIGH log lines**. The agent (Claude) **engaged with none of them during the session**, then — when asked about the tool — incorrectly claimed it "didn't run." Real, persistent findings were buried in the transient noise and only surfaced when the human manually pasted one and pushed back. Examples of genuine HIGHs that were missed until manual intervention:

- `onSubmit` awaited an RTK Query mutation without `.unwrap()` → failures treated as success, user navigated forward as if saved.
- A delete handler swallowed `.unwrap().catch(...)` then `navigation.replace(...)` unconditionally → failed deletes returned the user to the list as if removed.
- A `Loading` screen marked onboarding complete even when the completion PUT failed.

All correct, all HIGH, all sat unaddressed because the channel is ignorable.

## Root cause (two flaws)

1. **No gate.** `PostToolUse` writes to stderr/log; nothing forces the agent to act. Passive feedback in a long session reliably gets ignored.
2. **Per-edit noise.** Reviewing after *every* edit floods the log with transient mid-edit compile states ("missing import", "undefined X", "not yet in SCREENS_TO_NAVIGATOR_MAP") that resolve by the final state (final typecheck was clean). These drown the persistent findings — ~250 HIGH lines, the large majority transient.

### Core: a `Stop`-hook gate on the final state

Claude Code `Stop` hooks can return `{"decision":"block","reason":"…"}`, which prevents the agent from ending the turn and feeds the reason back as a must-address. The per-edit hook stays advisory; the **Stop hook is the gate**:

```
turn ends → Stop hook:
  git diff --name-only                       # files changed this turn
  for each file: read the LATEST log.jsonl entry  # = its final-state review, no new Codex call
  collect concerns.high not matched by an ack
  none → allow stop
  any  → decision:block "N unaddressed HIGH on X,Y: …  Fix or /codex-pair-ack <id> \"reason\"."
```

Reading the **latest** entry per changed file is the trick: it's the final-state review the per-edit hook already computed, so transient mid-edit findings are gone — accurate **and** zero extra Codex cost.

### Supporting pieces (ranked)

1. **`blockOn: HIGH` frontmatter** in `.codex-pair/context.md` — opt-in severity gate (HIGH blocks; MED/LOW advisory). The single smallest change that would have prevented the "all green" claim above.
2. **Ack protocol** (`/codex-pair-ack <id> "<reason>"`) — lets the agent *explicitly defer* a finding (stale / pre-existing / out-of-scope) with a recorded, auditable reason instead of silently ignoring. Keeps the gate humane (a pure block-on-HIGH would trap on a stale finding).
3. **Open-findings ledger** (`.codex-pair/open.json`) with auto-close on reconcile — when a later review of a file no longer reports a finding, close it. The gate reads *open* findings, not raw history. This structurally separates transient noise from real issues.
4. **Edit-significance gating** at the source — skip pure i18n/import/whitespace/rename edits. Fewer reviews, lower cost, far fewer transient findings.
5. **Debounce to quiescence** — review a file when it goes idle / at turn-end, not after every edit; naturally captures the final state.

## MVP

A `Stop` hook (~40 lines bash/jq): `git diff --name-only` → latest `log.jsonl` entry per file → block if any `concerns.high` without a matching ack. Everything else (ledger, debounce, significance-gating) is refinement on top.

## Why this granularity

Per-edit review maximizes **recall** (catches transient breakage too) but its output is unusable as a gate. Per-turn enforcement on the **latest** finding per file gives the recall benefit during editing **and** an accurate gate at the boundary — with no extra Codex calls, by reusing reviews already logged. Recall where you want it, precision where you enforce it.

---
*Filed from a real rn-dev session where the advisory channel was ignored end-to-end; this is the field replication that motivates the gate.*

### PR DESCRIPTION
**Pull request by Lykhoyda**
Closes #142.

## What

An **opt-in `Stop` hook** that blocks the agent from ending a turn while unaddressed **HIGH** codex-pair findings remain — making them un-ignorable — with **zero new LLM calls** (it reads the already-computed `.codex-pair/log.jsonl`). Deferrable per-finding via `/codex-pair-ack <hash> "<reason>"`.

Re-introduces a `Stop` hook that **ADR-048 removed**, but on a fundamentally different mechanism (see ADR-118 + the why-not-ADR-048 table in the design doc).

## Design highlights

- **Opt-in, default OFF** via `blockOn: HIGH` in `.codex-pair/context.md` — no surprise blocking for existing users.
- **Log-derived, reconciled against present reality** (the log is an append-only memory of edits, not a live view): `[A]` skip files not on disk (deletion/rename), `[C]` indeterminate latest verdict → fail-open for that file, `[B]` skip files clean-vs-HEAD via `git status --porcelain` (revert/branch-switch; additive filter, *not* ADR-048's scoping), `[E]` file-scoped content-hash acks (`hashConcernBody(relPath + ":" + text)`, shared with ADR-096).
- **Fail-open & loud** — any hook error warns to stderr and exits 0; a gate that blocks turns on its own bug is worse than no gate.
- Pure logic in `scripts/lib/stop-gate.mjs` (I/O-injected, **12 unit tests**); the hook is thin wiring.

## Antigravity review (read-only) — applied before this PR

The design was refined by an external Antigravity critique (which surfaced the A/B/C/E edge cases), and the **implementation** was then reviewed by Antigravity (verdict: SHIP WITH CHANGES). All findings addressed in the final commit:

- **git `execFileSync` had no timeout** → added `timeout: 5000` (a hung git would otherwise wedge turn-end).
- **macOS `/var`→`/private/var` realpath mismatch** → canonicalize `markerDir` + log file paths via `realpathSync` so they align with git's realpath'd repo root (fixes `[B]` filter + `[E]` ack hash on symlinked paths).
- **`/codex-pair-ack` shell-injection** → reason now passed via a single-quoted heredoc (not a double-quoted argv); hash validated as 16-char hex.
- **`retried`/`broker_fallback` lack a `file` field** → documented (already excluded by `selectLatestEntries`).

## Testing

- `@ask-llm/plugin`: **385 tests pass** (12 new stop-gate unit tests + manifest + skills registration).
- Hook smoke-tested both paths: allow (exit 0, no output) and block (emits `{"decision":"block",...}`).
- `yarn lint` + `yarn build` green. `@ask-llm/plugin` is `private` → no changeset.

## Out of scope (follow-up ADRs)

`open.json` ledger, edit-significance gating, debounce, MED/LOW gating, default-ON posture, concern-text normalization for LLM wording drift, `--no-cache` ack-clear.

Spec: `docs/plans/2026-06-09-codex-pair-stop-gate-design.md` · Plan: `docs/plans/2026-06-09-codex-pair-stop-gate-plan.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/claude-plugin/scripts/lib/stop-gate.mjs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+//   gitDirty  Set<absPath> | null          (null = no git filter)
+//   markerDir project root for relPath ack identity
+export function collectBlockingHighs({ entries, acks, existsFn, gitDirty, markerDir }) {
+  const blocking = [];
+  for (const [file, entry] of entries) {
+    if (!existsFn(file)) continue; // [A] deleted/renamed
+    if (INDETERMINATE.has(entry.verdict)) continue; // [C] indeterminate latest → fail-open
+    if (gitDirty && !gitDirty.has(file)) continue; // [B] clean vs HEAD
+    const highs = entry.concerns?.high ?? [];
+    for (const text of highs) {
```
[the remark this answers is not in the document]
Comment (Lykhoyda) [reply packages/claude-plugin/scripts/lib/stop-gate.mjs:48]: Fixed in 00598fb — `collectBlockingHighs` now uses `Array.isArray(entry.concerns?.high) ? … : []`, so a malformed non-array `high` can't iterate-as-string or throw.

File: packages/claude-plugin/scripts/lib/stop-gate.mjs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  const lines = blocking.map((b) => {
+    const rel = relative(markerDir, b.file);
+    return `  [${b.hash}] ${rel}\n    ${b.text}`;
+  });
+  return (
+    `🚫 codex-pair: ${blocking.length} unaddressed HIGH finding(s) (blockOn: HIGH). ` +
+    `Fix them, or defer each with /codex-pair-ack.\n\n` +
+    `${lines.join("\n")}\n\n` +
+    `To defer (stale / pre-existing / out-of-scope):\n` +
+    `  /codex-pair-ack ${blocking[0].hash} "<reason>"\n` +
```
[the remark this answers is not in the document]
Comment (Lykhoyda) [reply packages/claude-plugin/scripts/lib/stop-gate.mjs:67]: Reworded in 00598fb: the deferral line now reads 'run /codex-pair-ack with its [hash] from the list above — e.g. …'. Each finding already prints its own `[hash]` in the list, so all are ackable; the e.g. just shows the form.

File: packages/claude-plugin/scripts/codex-pair-stop-gate.mjs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+// inside the leading `---` block.
+function readBlockOn(markerDir) {
+  let text;
+  try {
+    text = readFileSync(contextPath(markerDir), "utf8");
+  } catch {
+    return null;
+  }
+  const fm = text.match(/^---\n([\s\S]*?)\n---/);
+  if (!fm) return null;
```
[the remark this answers is not in the document]
Comment (Lykhoyda) [reply packages/claude-plugin/scripts/codex-pair-stop-gate.mjs:49]: Fixed in 00598fb — the frontmatter regex is now `/^---\r?\n([\s\S]*?)\r?\n---/`, so a CRLF-edited context.md no longer silently disables the gate on Windows.

File: packages/claude-plugin/scripts/lib/state.mjs [hunk 4]
```diff
@@ -75,6 +75,25 @@ export const cacheRoot = (markerDir) => join(pairRoot(markerDir), CACHE_DIR);
...
+export function readAcks(markerDir) {
+  try {
+    return JSON.parse(readFileSync(acksPath(markerDir), "utf8"));
+  } catch {
+    return {}; // missing/corrupt → no acks
+  }
+}
+
+// Append-merge a single ack. Best-effort; the slash command surfaces failures.
+export function addAck(markerDir, hash, { reason }) {
```
[the remark this answers is not in the document]
Comment (Lykhoyda) [reply packages/claude-plugin/scripts/lib/state.mjs:93]: Fixed in 00598fb — `addAck` now writes to `acks.json.tmp` then `renameSync`s into place (atomic), so a kill mid-write can't corrupt/truncate the store.

File: packages/claude-plugin/skills/codex-pair-ack/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+   codex-pair is not enabled in this project (no .codex-pair/context.md found). Nothing to acknowledge.
+   ```
+
+3. **Validate the hash** before using it: it must match `^[0-9a-f]{16}$`. If it does not, print the usage block from step 1 and stop (a malformed hash means the user mistyped it).
+
+4. **Record the ack** by running the Bash command below. Substitute real values for `<plugin-root>` (the absolute path of this plugin, i.e. the value of `CLAUDE_PLUGIN_ROOT`), `<markerDir>`, and `<hash>`. **Do NOT interpolate `<reason>` into the command string** — the reason is read from stdin via a single-quoted heredoc so the shell never expands `$(...)`, backticks, or quotes inside it (injection-safe). Replace the `<reason>` line with the user's reason text verbatim:
+   ```bash
+   node --input-type=module -e '
+   import { addAck } from "<plugin-root>/scripts/lib/state.mjs";
+   import { readFileSync } from "node:fs";
```
[the remark this answers is not in the document]
Comment (Lykhoyda) [reply packages/claude-plugin/skills/codex-pair-ack/SKILL.md:30]: Verified empirically — this is **not** off-by-one for `-e`. `node --input-type=module -e 'CODE' A B` yields `process.argv = [execPath, "A", "B"]` (no script-file placeholder at argv[1]; that only happens when running a script *file*). Tested on the project's Node (v24.13.0, this exact invocation):
```
$ node --input-type=module -e 'console.log(JSON.stringify(process.argv))' MARKERDIR HASHVAL
[".../node","MARKERDIR","HASHVAL"]
```
So `argv[1]=markerDir`, `argv[2]=hash` is correct; the suggested `argv[2]`/`argv[3]` would actually break it (hash→markerDir, undefined→hash). Kept the code and added a clarifying comment.
